qiita_db/analysis.py

```diff
@@ -808,81 +808,113 @@ def build_files(self,
             # different stages and possible errors.
             samples = self.samples
+
             # figuring out if we are going to have duplicated samples, again
             # doing it here cause it's computational cheaper
-            all_ids = list(chain.from_iterable([
-                samps for _, samps in viewitems(samples)]))
-            rename_dup_samples = ((len(all_ids) != len(set(all_ids))) and
-                                  merge_duplicated_sample_ids)
+            # 1. merge samples per data_type, reference used and the command id
+            #    if not one of the tables is not 'Pick closed-reference OTUs'
+            #    it's safer to always rename. Note that grouped_samples is
+            #    basically how many biom tables we are going to create
+            rename_dup_samples = False
+            grouped_samples = {}
+            for k, v in viewitems(samples):
+                a = qdb.artifact.Artifact(k)
+                p = a.processing_parameters
+                c = p.command
+                if c.name != 'Pick closed-reference OTUs':
+                    rename_dup_samples = True
+                    break
+                l = "%s.%d.%d" % (a.data_type, p.values['reference'], c.id)
+                if l not in grouped_samples:
+                    grouped_samples[l] = []
+                grouped_samples[l].append((k, v))
+            # 2. if rename_dup_samples is still False, make sure that we don't
+            #    need to rename samples by checking that there are not
+            #    duplicated samples per group
+            if not rename_dup_samples:
+                for k, v in viewitems(grouped_samples):
+                    # this element only has one table, continue
+                    if len(v) == 1:
+                        continue
+                    dup_samples = set()
+                    for k, v in v:
+                        v = set(v)
+                        if dup_samples & v:
+                            rename_dup_samples = (True and
+                                                  merge_duplicated_sample_ids)
+                            break
+                        dup_samples = dup_samples | v
 
             self._build_mapping_file(samples, rename_dup_samples)
-            self._build_biom_tables(samples, rarefaction_depth,
+            self._build_biom_tables(grouped_samples, rarefaction_depth,
                                     rename_dup_samples)
 
-    def _build_biom_tables(self, samples, rarefaction_depth=None,
+    def _build_biom_tables(self, grouped_samples, rarefaction_depth=None,
                            rename_dup_samples=False):
         """Build tables and add them to the analysis"""
         with qdb.sql_connection.TRN:
             base_fp = qdb.util.get_work_base_dir()
 
-            # this assumes that there is only one reference/pipeline for each
-            # data_type issue #164
```

Comment:
Since it seems like this issue will be closed, can you make a comment in an upcoming commit message (if there's one) to close this issue automatically?